### DISCUSSION TITLE
Support for WS2812 Cubes

### DISCUSSION THREAD

**Discussion by zango-me** [PlummersSoftwareLLC/NightDriverStrip#9] [Ideas]
First of all: Awesome project and thanks for the very cool youtube videos Dave!

I did a lot with WS2812 LEDs in the past and have a lot of custom lights deployed in my house at the moment.
I'm currently running WLED to drive them but I'm looking into switchung to this in the future in hope of better performance while syncronizing them with live pixel data from a centralized node/system (WLED glitches out a bit with too many LEDs/devices currently) + the microphone support sounds very intriguing.

What would be a real advantage over any other solution I know currently exists tho would be the support for WS2812 Cubes (like this one from led-genial https://www.led-genial.de/Polonium-Cube-F2-der-LED-Wuerfel-Bausatz - these panels can also be found for cheap on aliexpress and ebay for a complete diy build)

I already got cubes in different sizes built but never found a solution to drive effects on them properly (because of the 3D structure and how the edges of the pannels connect to each other the effects just don't look good)

maybe this might be something worth looking into to support at some point in the future! :)

![photo_2021-10-03_23-45-22](https://user-images.githubusercontent.com/5586490/135772533-ef7aba2e-2d7b-4304-88dc-4ab862936856.jpg)

**Comment by davepl** [comment]
Cool!  What if I simply added an overload of the setPixel() API that allowed for a Z component, where Z was the cube face?  The thing is someone who owns cubes would have to then write effects specially tailored for the cubes.  If you can convince a vendor to send me som cubes, I'd do it myself for fun!  But I don't own any currently...

**Reply by zango-me** [reply]
I can send you a set of 6 8x8 panels if you like :)

**Comment by robertlipe** [comment]
Interesting. I was talking to someone about cubes of 16x16 faces and the reality that 6*16*16*3*256 approached "a lot" pretty quickly. The thought of adding another wire (when they were already having to add several for power just for injection) was a deal-breaker and their project just fell apart.

Were these ever produced and sold in quantity?

Are you perhaps using the [ECO leds](https://www.lcsc.com/datasheet/lcsc_datasheet_1811151230_Worldsemi-WS2812E_C139127.pdf
) to help with the power consumption/heat dissipation issues?

Were there special effects developed for them or did you just run planar effects projected onto a larger area and just subset them? If you build code for a 16x16 panel, you're likely to get 4 faces that have somewhat natural-ish edges and 2 faces that have harsh transitions. (Of course, math being what it is, those two faces may make more than two visible surface look funky.)

Did you attempt to get past the frame rate issues of the 16x16 via multiple data lines or did you just rely on 8x8 panels inherently downscaling it and making that better?

Similarly, one of the demos shows it demonstrated on a flat plane so that four surfaces (where have we heard that number before? :-) ) might look good and then just ignoring the top and bottom winee they're no projected into the room?

I can imagine some of the more geometric-y, math-y, (boring) effects from "my" effects branch actually working pretty well on these in the way that, say, a GIF player or a clock would just turn into a mess. Are there custom effects that work well on these? My read is that the current 'cube' build in ND is for something else.

Were there effects that WLED had that we should study for potential inspiration?

I'm interested in knowing what it'd take to move this to an "answered" state where we support interesting devices like this really well. (I'm getting bored with plain ole strips.)
